### ISSUE TITLE
Slow modelio namelist generation when NINST is large

### ISSUE THREAD

**Issue by kdraeder** [ESMCI/cime#4383]
Data assimilation (DA) jobs require 20 < NINST < 1000
and typically run multiple cycles (calls to ESP) in each job.
Each cycle requires a new run of CESM, which prompts calls to preview_namelist
and the generation of namelists for all of the instances of all of the components,
serially (while a large number of nodes needed for the job sit idle).
The namelists are either identical from cycle to cycle, 
or have minor changes such as time stamps in file names.
The ${component}\_in\_${inst} creation can be turned off by 
   `case.submit --skip-preview-namelist`
but the modelio namelists can't.

We've modified 
   `src/drivers/mct/cime_config/buildnml`
to check for the existence of a modelio namelist.
If it exists, merely replace some of the contents with new data,
otherwise, create the namelist as usual.
It reduced the time required to write the modelio files for 80 instances
from 40 s to 0.6 s, out of a total cycle time of \~400 s.
I've attached a modified version (with .txt added so that github will accept it), 
from a cime branch from origin/maint-5.6  (\~ Mar 12 05:20:53 2019)
170db73e00130294bac98fde5dbbcdcde06a33d6

> \> git describe --tags
>     cime5.6.11-87-g0a8bce9f1
> 

Tell me if I need to import this into a more recent CIME in order
for it to be reasonable to work with.
I understand that this will need to be implemented in NUOPC/CMEPS.

This may not be the solution that's optimal for CIME as a whole, 
but it was sufficient for our use and has a small footprint.
I'll submit it as a pull request, if it looks like the right strategy.
[buildnml.txt](https://github.com/ESMCI/cime/files/11114117/buildnml.txt)

**Comment by jedwards4b** [comment]
Yes, please submit this as a pull request against the maint-5.6 branch of cime.
The mct driver is deprecated and will not be supported in cesm3, we will need to address similar issues in the cmeps code base separately.

**Comment by kdraeder** [comment]
My old branch from maint-5.6 has all of the changes I made to run a large assimilation project.
I want to merge only some of those into maint-5.6, in small batches.
So I've cloned a new CESM+CIME: cesm2_3_beta03, checked out the maint-5.6 branch,
and created a new branch from that (m5.6_dart_integ).  
Then I imported the changes to src/drivers/mct/cime_config/buildnml and committed them.
So they haven't been tested in this context.

I'm not sure what to do next.  Should I push this (branch?) to the remote (https://github.com/ESMCI/cime)
and then issue a PR that compares that branch to the maint-5.6 branch?

**Comment by jedwards4b** [comment]
cesm2_3_* is not compatible with the maint-5.6 branch of cime.  If you want to test your changes in a functional cesm environment the place to start is the 
cesm2.1-alphabranch.

That said it looks like you followed the correct steps for the cime repo.

The preferred way to submit a PR to cime is for you to create a fork of the repository and push your branch to the fork, then open a PR in ESMCI/CIME with the incoming pointing to the branch on your fork.

**Comment by kdraeder** [comment]
The upgrade in maint-5.6 has been merged into a new cesm2.1.4-rc.12 tag.

It appears that the upgrade does not need to be propagated forward to a recent CESM+CIME.
Under nuopc and cmeps the modelio files don't appear in the run directory
and @jedwards4b says that they are handled by fortran code.
So I'll assume that they're handled in parallel and reasonably fast, until I see otherwise.
